strawberryfields/engine.py

```diff
@@ -609,15 +610,63 @@ def run_async(self, program: Program, *, compile_options=None, **kwargs) -> Job:
         """
         # get the specific chip to submit the program to
         compile_options = compile_options or {}
+        recompile = compile_options.pop("recompile", False)
         kwargs.update(self._backend_options)
 
         device = self.device_spec
 
         compiler_name = compile_options.get("compiler", device.default_compiler)
-        msg = f"Compiling program for device {device.target} using compiler {compiler_name}."
-        self.log.info(msg)
 
-        program = program.compile(device=device, **compile_options)
+        if program.compile_info is None:
+            # program is uncompiled
+            msg = f"Compiling program for device {device.target} using compiler {compiler_name}."
+            self.log.info(msg)
+            program = program.compile(device=device, **compile_options)
+
+        elif (
+            program.compile_info[0].target != device.target
+            or program.compile_info[0]._spec != device._spec
+        ):
+            # program was compiled for a different device
+            if not recompile:
+                raise ValueError(
+                    "Cannot use program compiled for "
+                    f"{program._compile_info[0].target} for target "
+                    '{self.target}. Set the "recompile" entry in the '
+                    "compile_options dictionary to true to compile with "
+                    "{compiler_name}."
+                )
+
+            msg = f"Recompiling program for device {device.target} using compiler {compiler_name}."
+            self.log.info(msg)
+            program = program.compile(device=device, **compile_options)
+
+        elif program.compile_info[1] != compiler_name:
+            # program was compiled for the device, but using a different compiler
+
+            if compile_options is None or not compile_options:
+                msg = (
+                    f"No compile options specified, compiling program for "
+                    f"device {device.target} using compiler Xstrict."
```

Comment:
Perhaps we can make this a bit clearer?
```suggestion
                    f"Program previously compiled for {device.target} using {program.compile_info[1]}. "
                    f"Validating program against the Xstrict compiler."
```